src/Npgsql/PostgresException.cs

```diff
@@ -293,22 +290,44 @@ public override IDictionary Data
         {
             get
             {
-                // Remarks: return Dictionary with object keys although all our keys are string keys
-                // because System.Windows.Threading.Dispatcher relies on that
-                return _data ?? (_data = (
-                    from p in typeof(PostgresException).GetProperties()
-                    let k = p.Name
-                    where p.Name != nameof(Data)
-                    where p.GetCustomAttribute<PublicAPIAttribute>() != null
-                    let v = p.GetValue(this)
-                    where v != null
-                    where k != nameof(Position) && k != nameof(InternalPosition) || (int)v != 0
-                    select new { Key = k, Value = v }
-                    ).ToDictionary(kv => (object)kv.Key, kv => kv.Value)
-                );
+                if (_data == null)
+                {
+                    _data = base.Data;
+                    AddData(nameof(Severity), Severity);
+                    AddData(nameof(SqlState), SqlState);
+                    AddData(nameof(MessageText), MessageText);
+                    AddData(nameof(Detail), Detail);
+                    AddData(nameof(Hint), Hint);
+                    AddData(nameof(Position), Position);
+                    AddData(nameof(InternalPosition), InternalPosition);
+                    AddData(nameof(InternalQuery), InternalQuery);
+                    AddData(nameof(Where), Where);
+                    AddData(nameof(SchemaName), SchemaName);
+                    AddData(nameof(TableName), TableName);
+                    AddData(nameof(ColumnName), ColumnName);
+                    AddData(nameof(DataTypeName), DataTypeName);
+                    AddData(nameof(ConstraintName), ConstraintName);
+                    AddData(nameof(File), File);
+                    AddData(nameof(Line), Line);
+                    AddData(nameof(Routine), Routine);
+                }
+
+                return _data;
             }
         }
 
+        void AddData(string key, int value)
```

Comment:
Are local functions allowed inside a getter? If so, you could move them there for better scoping.

Reply:
Allowed, but they can't be resolved correctly if they have the same name.

Reply:
I hadn't considered that, makes sense!

Reply:
Done.